### PR TITLE
[RLlib] Sub-class `Trainer` (instead of `build_trainer()`): All remaining classes; soft-deprecate `build_trainer`.

### PR DESCRIPTION
**Pull request by sven1977**
Sub-class `Trainer` (instead of `build_trainer()`): All remaining classes; soft-deprecate `build_trainer`.
- MADDPG
- AlphaZero
- Bandits
- BC/MARWIL
- Dreamer
- SlateQ

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/dqn/simple_q.py [hunk 1]
```diff
@@ -112,35 +112,59 @@
 # yapf: enable
 
 
-def get_policy_class(config: TrainerConfigDict) -> Optional[Type[Policy]]:
-    """Policy class picker function. Class is chosen based on DL-framework.
+# TODO: Move into SimpleQ once all OffPolicyTrainers have been converted to
```
Comment (gjoliver) [inline rllib/agents/dqn/simple_q.py:115]: still applies?
if these are the last remaining classes, why don't move this and execution_plan() below into SimpleQ class?

File: rllib/agents/dqn/simple_q.py [hunk 2]
```diff
@@ -112,35 +112,59 @@
...
+    # Multi-agent mode and multi-GPU optimizer.
+    if config["multiagent"]["policies"] and \
+            not config["simple_optimizer"]:
+        logger.info(
+            "In multi-agent mode, policies will be optimized sequentially"
+            " by the multi-GPU optimizer. Consider setting "
+            "simple_optimizer=True if this doesn't work for you.")
+
+
+# TODO: Move into SimpleQ once all OffPolicyTrainers have been converted to
```
Comment (gjoliver) [inline rllib/agents/dqn/simple_q.py:162]: same here.

**Review by gjoliver — COMMENTED** [review]
1 really quick question

File: rllib/agents/dqn/simple_q.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/dqn/simple_q.py]: Good catch, this PR should deprecate `OffPolicyTrainer` entirely.

**Review by gjoliver — APPROVED** [review]
Thanks \o/ :)

**Pull request merged**